File: android/src/main/java/com/ionic/plugin/android/capacitor/core/actions/CallContext.kt
package com.ionic.plugin.android.capacitor.core.actions

import com.getcapacitor.PluginCall
import com.ionic.plugin.android.capacitor.core.WrapperDelegate
import com.ionic.plugin.android.capacitor.core.toJSObject
import com.ionic.plugin.android.core.actions.CallContext
import com.ionic.plugin.core.PluginException
import com.ionic.plugin.core.PluginExceptionBase
import com.ionic.plugin.core.actions.Mappers
import com.spryrocks.kson.JsonArray
import com.spryrocks.kson.JsonObject

class CallContext(
    private val call: PluginCall,
    wrapperDelegate: WrapperDelegate,
    private val mappers: Mappers,
) : CallContext(wrapperDelegate) {
    override fun asObject() = AsObject(call)

    override fun asArray(): AsArray {
        throw PluginException("Not implemented")
    }

    class AsObject(private val call: PluginCall) :
        com.ionic.plugin.core.actions.CallContext.AsObject() {
        override fun opt_(name: String) = throw NotImplementedError()
        override fun optString_(name: String) = wrapNullable(name) { call.getString(name) }
        override fun optInt_(name: String) = wrapNullable(name) { call.getInt(name) }
        override fun optBoolean_(name: String) = wrapNullable(name) { call.getBoolean(name) }
        override fun optFloat_(name: String) = wrapNullable(name) { call.getFloat(name) }
        override fun optDouble_(name: String) = wrapNullable(name) { call.getDouble(name) }
        override fun optJsonObject_(name: String) = wrapNullable(name) {
            val jsonString = call.getObject(name).toString()
            return@wrapNullable JsonObject.fromJson(jsonString)
        }
        override fun optLong_(name: String) = wrapNullable(name) { call.getInt(name)?.toLong() }
        override fun optNumber_(name: String) = wrapNullable(name) {
            try {
                return@wrapNullable call.getDouble(name)
            } catch (e: Throwable) {
                return@wrapNullable call.getInt(name) as Number
            }
        }
        override fun optJsonArray_(name: String) = wrapNullable(name) {
            val jsonString = call.getArray(name).toString()
            return@wrapNullable JsonArray.fromJson(jsonString)
        }

        private fun <T> nullable(key: String, getter: () -> T?): T? {
            val data = call.data
            if (!data.has(key)) return null
            if (data.isNull(key)) return null
            return getter()
        }

      private fun <T> wrapNullable(key: String, getter: () -> T?): ValueWrapper<T> {
        val value = nullable(key, getter)
        return ValueWrapper(value)
      }
    }

    override fun success(data: Any?, finish: Boolean) {
        when (data) {
            null -> resolve(finish)
            is JsonObject -> resolve(finish, data)
            else -> throw NotImplementedError("This data type is not supported")
        }
    }

    override fun error(error: Throwable?, finish: Boolean) {
      val exception: Exception? = when (error) {
        is Exception -> error
        is Throwable -> Exception(error)
        else -> null
      }

      val defaultMessage = "Unknown error"

      val message = exception?.message ?: defaultMessage
      val data = (exception as? PluginExceptionBase)?.let(mappers.errorMapper::mapToJson)

      reject(finish, message, data)
    }

    private fun resolve(finish: Boolean, data: JsonObject? = null) {
        setFinish(finish)
        if (data != null) {
            call.resolve(data.toJSObject())
        } else {
            call.resolve()
        }
    }

    private fun reject(finish: Boolean, message: String, data: JsonObject?) {
        setFinish(finish)
        if (data != null) {
            call.reject(message, data.toString())
        } else {
            call.reject(message)
        }
    }

    private fun setFinish(finish: Boolean) {
      if (!finish) call.save()
    }
}

File: ios/src/actions/CallContext.swift
import Capacitor

public class CallContext {
    private let call: CAPPluginCall
    private let mappers: CoreMappers

    init(call: CAPPluginCall, mappers: CoreMappers) {
        self.call = call
        self.mappers = mappers
    }

    public func asObject() -> CallContextAsJsonObject {
        return CallContextAsJsonObject(call: call)
    }

    public func success(_ data: JsonObject?, finish: Bool) {
        call.keepAlive = !finish
        
        if let data = data {
            call.resolve(data.toRaw() as PluginCallResultData)
        } else {
            call.resolve()
        }
    }

    public func error(_ error: Error?, finish: Bool) {
        call.keepAlive = !finish

        let errorData = prepareErrorData(error)

        let message = error?.localizedDescription ?? "Unknown error"
        
        if let errorData = errorData {
            call.reject(message, nil, nil, errorData.toRaw() as PluginCallResultData)
        } else {
            call.reject(message)
        }
    }

    private func prepareErrorData(_ error: Error?) -> JsonObject? {
        guard let error = error else {
            return nil
        }
        let pluginError = mappers.errorMapper.map(error)
        return mappers.errorMapper.mapToJson(pluginError)
    }
}

public class CallContextAsJsonObject: IJsonObjectProperties {
    private let call: CAPPluginCall

    init(call: CAPPluginCall) {
        self.call = call
    }

    public func opt(_ name: String) -> JsonValue? {
        return nil
    }

    public func optString(_ name: String) -> String? {
        return call.getString(name)
    }

    public func optInt(_ name: String) -> Int? {
        return call.getInt(name)
    }

    public func optDouble(_ name: String) -> Double? {
        return call.getDouble(name)
    }

    public func optBool(_ name: String) -> Bool? {
        return call.getBool(name)
    }

    public func optObject(_ name: String) -> JsonObject? {
        guard let raw = call.getObject(name)
        else { return nil }
        return JsonObject.fromRaw(raw)
    }

    public func optArray(_ name: String) -> JsonArray? {
        guard let raw = call.getArray(name)
        else { return nil }
        return JsonArray.fromRaw(raw)
    }

    public func get(_ name: String) throws -> JsonValue {
        guard let result = opt(name) else {
            throw PluginError(message: "Value for required '\(name)' is nil")
        }
        return result
    }

    public func getString(_ name: String) throws -> String {
        guard let result = optString(name) else {
            throw PluginError(message: "Value for required string '\(name)' is nil")
        }
        return result
    }

    public func getInt(_ name: String) throws -> Int {
        guard let result = optInt(name) else {
            throw PluginError(message: "Value for required int '\(name)' is nil")
        }
        return result
    }

    public func getDouble(_ name: String) throws -> Double {
        guard let result = optDouble(name) else {
            throw PluginError(message: "Value for required double '\(name)' is nil")
        }
        return result
    }

    public func getBool(_ name: String) throws -> Bool {
        guard let result = optBool(name) else {
            throw PluginError(message: "Value for required bool '\(name)' is nil")
        }
        return result
    }

    public func getObject(_ name: String) throws -> JsonObject {
        guard let result = optObject(name) else {
            throw PluginError(message: "Value for required object '\(name)' is nil")
        }
        return result
    }

    public func getArray(_ name: String) throws -> JsonArray {
        guard let result = optArray(name) else {
            throw PluginError(message: "Value for required array '\(name)' is nil")
        }
        return result
    }
}
